sdk/python/kfp/v2/dsl/component_bridge.py

```diff
@@ -0,0 +1,199 @@
+      raise TypeError(
+          'ContainerOp object {} was passed to component as an input argument. '
+          'Pass a single output instead.'.format(input_name))
+    elif isinstance(argument_value, dsl.PipelineParam):
+      input_type = component_spec._inputs_dict[input_name].type
+      reference_type = argument_value.param_type
+      types.verify_type_compatibility(
+          reference_type, input_type,
+          'Incompatible argument passed to the input "{}" of component "{}": '
+          .format(input_name, component_spec.name))
+
+      arguments[input_name] = str(argument_value)
+
+      if type_utils.is_artifact_type(input_type):
+        # argument_value.op_name could be none, in which case an importer node
+        # will be inserted later. Use output_artifact_key to preserve the name
+        # of pipeline parameter which is needed by importer.
+        pipeline_task_spec.inputs.artifacts[input_name].producer_task = (
+            argument_value.op_name or '')
+        pipeline_task_spec.inputs.artifacts[input_name].output_artifact_key = (
+            argument_value.name)
+      elif type_utils.is_parameter_type(input_type):
+        pipeline_task_spec.inputs.parameters[
+            input_name].runtime_value.runtime_parameter = argument_value.name
+      else:
+        raise NotImplementedError(
+            'Unsupported input type: "{}"'.format(input_type))
+    elif isinstance(argument_value, str):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.string_value = argument_value
+    elif isinstance(argument_value, int):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.int_value = argument_value
+    elif isinstance(argument_value, float):
+      pipeline_task_spec.inputs.parameters[
+          input_name].runtime_value.constant_value.double_value = argument_value
+    else:
+      raise NotImplementedError(
+          'Input argument supports only the following types: PipelineParam'
+          ', str, int, float. Got: "{}".'.format(argument_value))
+
+  for output in component_spec.outputs or []:
+    normalized_output_type = output.type.lower()
+    if type_utils.is_artifact_type(output.type):
+      pipeline_task_spec.outputs.artifacts[
+          output
+          .name].artifact_type.schema_title = type_utils.get_artifact_type(
+              output.type)
+    elif type_utils.is_parameter_type(output.type):
+      pipeline_task_spec.outputs.parameters[
+          output.name].type = type_utils.get_parameter_type(output.type)
+    else:
+      raise NotImplementedError('Unsupported output type: "{}"'.format(
+          output.type))
+
+  def _input_artifact_placeholder(input_key: str) -> str:
+    return "{{{{$.inputs.artifacts['{}'].path}}}}".format(input_key)
+
+  def _input_parameter_placeholder(input_key: str) -> str:
+    return "{{{{$.inputs.parameters['{}']}}}}".format(input_key)
```

Comment:
I'm not sure `inputs.parameters` will work. It might be better to use `{{$.inputs.artifacts['{}'].value}}`.

Reply:
`inputs.parameters` works for primitive types including string, int, and double.
When did we introduce `artifacts[].value`? and how is it different from `input.parameter`?

Reply:
> When did we introduce artifacts[].value?

This is along the line of how primitive-typed artifact is being passed in TFX. That is, the input is still an artifact, but the component can access its raw content.

And `input.parameters` here is an analogy to the `exec_property` in TFX.

@Ark-kun  I think `input.parameters` should be a first-class citizen for the pipeline IR. Do you mind providing a counter example where it does not work? Thanks!

Reply:
> @Ark-kun I think `input.parameters` should be a first-class citizen for the pipeline IR. Do you mind providing a counter example where it does not work? Thanks!

Currently there is only one kind of inputs and outputs and they can be freely connected (given that the types are compatible). Fragmenting the inputs and outputs into artifacts and parameters breaks all cases where connected inputs and outputs are now discriminated into different categories.

Some of the use cases that are going to break:

Using a uploader component to upload a string or a numerical value.

Using a downloader component to download a string that's later passed to come component.

A component that queries a JSON object. The output of such component can currently be used in any way.

Reply:
I think in order to make existing KFP components not break, we'll need to do the translation based on how things connect, i.e., look at how the component consumes some output. Here are the possibilities:

{ `PipelineParameter`, `OutputPath`} -> { `InputPath`, `InputValue` }

Let's assume notation \<ComponentName\>[\<Type\>] below with `->` indicating how things are wired up in the pipeline. The possibilities are:

1. `PipelineParameter` - > `C1[InputPath]`
    * Implies `C1` has an input `Artifact`
    * The input artifact must be created by an implicit ImporterNode to register in MLMD
    * The placeholder for `C1` should be `inputs.artifacts['..'].path`

2. `PipelineParameter` - > `C1[InputValue]`
    * Implies `C1` has an input `Parameter` that the system can store
    * The placeholder for `C1` should be `inputs.parameters['..']`

3. `PipelineParameter` - > `C1[InputValue]` **and** `PipelineParameter` - > `C2[InputPath]`
    * Implies `C1` and `C2` have an input `Artifact`
    * The input artifact must be created by an implicit ImporterNode to register in MLMD
    * The placeholder for `C1` should be `inputs.artifacts['..'].value`
    * The placeholder for `C2` should be `inputs.artifacts['..'].path`

4. `C0[OutputPath]` - > `C1[InputPath]`
    * Implies `C0` has an output `Artifact`
    * Implies `C1` has an input `Artifact`
    * The placeholder for `C0` should be `outputs.artifacts['..'].path`
    * The placeholder for `C1` should be `inputs.artifacts['..'].path`

5. `C0[OutputPath]` - > `C1[InputValue]`
    * Implies `C0` has an output `Parameter`
    * Implies `C1` has an input `Parameter`
    * The placeholder for `C0` should be `outputs.parameters['..'].output_file`
    * The placeholder for `C1` should be `inputs.parameters['..']`

6. `C0[OutputPath]` - > `C1[InputValue]` **and** `C0[OutputPath]` - > `C2[InputPath]`
    * Implies `C0` has an output `Artifact`
    * Implies `C1` and `C2` have an input `Artifact`
    * The placeholder for `C0` should be `outputs.artifacts['..'].path`
    * The placeholder for `C1` should be `inputs.artifacts['..'].value`
    * The placeholder for `C2` should be `inputs.artifacts['..'].path`


Note that in none of the above have we ended up using `.uri`. I believe there are two ways we can approach this:
* introduce `InputUri` placeholder in KFP
* switch on `.path` vs `.uri` based on KFP component annotation in the YAML

Finally, also note that we did not have to look at the type annotation (e.g. `GCSPath`) yet.

/cc @james-jwu 
/cc @paveldournov 
/cc @hongye-sun 

Thoughts?

Reply:
Thanks @neuromage ! By enumerating all the possibilities I think some potential caveats are manifested... For example

> 3. `PipelineParameter - > C1[InputValue]` and `PipelineParameter - > C2[InputPath]`

IIUC when we do this, the uri of `C2[InputPath]` artifact will be specified by the pipeline parameter. Do we consider `C1[InputValue]` and `C2[InputPath]` to be the same artifact and imported by the same importer? If so, then I think semantically we'll have `artifact.path == artifact.value` which is a bit of weird... How about we make the input for `C1` parameter in this case?

Reply:
Yes, we will consider them to be the same artifact imported by the same importer.

> If so, then I think semantically we'll have `artifact.path == artifact.value`
You're right, the placeholder for `C1` in this case should be `inputs.artifacts.path` instead of `.value`! Let me fix this in-place.

Reply:
Thank you for summarising the cases and explicitly outlining the behavior. I especially liked that you've mentioned the mixed consumption cases where the same piece of data is consumed in different ways.
I think this solves most of the compatibility issues.

I'd like to highlight possible issues in couple of cases.

#### Cases 1 and 3
>Case 1. Implies C1 has an input Artifact, whose URI is specified as the value of the PipelineParameter
>Case 3. Implies C1 and C2 have an input Artifact, whose URI is specified as the value of the PipelineParameter

I'm not sure that "PipelineParam is consumed using `inputPath`" implies "PipelineParam value is a URI" in most cases. In KFP I'd expect it to almost never be the case (URI-type inputs are in ~100% cases consumed using `inputValue`). So something consumed using `inputPath` is almost definitely not a URI.

What are the common cases where a pipeline parameter or a constant value is passed component that consumes it using `inputPath`?
It could be that the pipeline parameter contains a proto message, JSON data or text, something that can be both big and small while still having the same type. And the downstream component C1 could just be consuming the data via inputPath because that is easier and allows it to support bigger input data.

How can we then make cases 1 and 3 possible (if we consider that PipelineParam is **not** a URI)? We could inject an uploader-like component that takes `Parameter` input and outputs it as an `Artifact`.

So the full solution to cases 1,2,3 could be:

* `PipelineParam` or constant value consumed using `inputValue` -> Use `inputs.parameters['..']`
* `PipelineParam` or constant value consumed using `inputPath` -> Insert uploader-like component that coverts it to an artifact, then replace `inputPath` with `inputs.artifacts['..'].path` as usual (cases 4 and 6).


#### Case 7.
I think there is a missing case 7 where `C0[OutputPath]` is not consumed at all. Perhaps in this case we can handle it the same as case 4 and consider `C0` to have an output `Artifact` and use `outputs.artifacts['..'].path`

#### Cases 5 and 6.

Case 5 (and 4) might be treated a sub-case of the more general case 6. If we just always follow case 6, the code would be simpler and the behavior might be easier to understand.

#### Case 3.

>If so, then I think semantically we'll have artifact.path == artifact.value which is a bit of weird... How about we make the input for C1 parameter in this case?

I have a similar thought.

>You're right, the placeholder for C1 in this case should be inputs.artifacts.path instead of .value! Let me fix this in-place.

I'm not sure that change has solved the issue that Jiaxiao has highlighted.

Let's think this scenario through (even though I question some of its implications):
The PipelineParam has a URI. This means that the type of that PipelineParam should be `Uri` (if typed). This PipelineParam is passed to component `C1` input. That means that the input should have the same `Uri` type (if typed).
As stated, the component `C1` implementation consumes that input argument as value (`inputValue`). The value of a Uri argument is the string representation of that URI. So the component code expects to see a URI string in the command-line, not a local path.
This is why I'm not sure `inputValue` should ever be resolved to `inputs.artifacts['..'].path`. `inputValue` can be resolved to `inputs.artifacts['..'].value` or `inputs.parameters['..']`. (I can vaguely imagine a case where `inputValue` can be resolved to `inputs.artifacts['..'].uri`, but the conditions would be pretty convoluted like when connecting output of type `T` to input of type `Uri: {data_type: T}`)

#### Where to put the information that the component code expects a URI - input spec vs. placeholder.

>Note that in none of the above have we ended up using .uri. I believe there are two ways we can approach this:
>* introduce InputUri placeholder in KFP
>* switch on .path vs .uri based on KFP component annotation in the YAML

I've written a small document about these options last week that I'd like to share with you.
The gist is that the choice of whether to put the some type-related information in the component interface (input specs) or implementation (placeholders) depends on the capabilities of the execution system.
If the system can automatically bridge components that produce and consume data differently (e.g. produce as a file, consume as a uri or as value and vice versa) then the distinction is an implementation detail and belongs to the implementation. Otherwise the distinction is important to the pipeline author and needs to be part of the component interface.
Since the current DSL compiler for Argo does not erase the distinction between Artifact URIs and Artifact values, we likely need to have the destinction in the input specifications. If we improve the Argo compiler to erase the distinction, then we can add the inputUri placeholder. The IR Artifacts erase theis destinction (since every artifact has `.uri`, `.path` and `.value`), so we're on the right track. But the IR->Argo compiler is not ready yet.

Reply:
Alternative simplified placeholder conversion flowchart:

* `outputPath`: Always replace with `outputs.artifacts['..'].path`
* `inputPath`: Always replace with `inputs.artifacts['..'].path`
* * If the input argument is a constant value or a pipeline parameter (so, not an artifact), then inject an uploader component that creates an artifact out of that constant or a pipeline parameter.
* `inputValue`:
* * If the input argument comes from upstream component, replace with `inputs.artifacts['..'].value`
* * Otherwise (the input argument is a constant value or a pipeline parameter) you can use `inputs.parameters['..']`. (Alternatively use the uploader technique  from `inputPath`)